Paywall

vctt94 opened the pull request:
fixes #106 

Still need some styling and a faucet button for automatic testing, as mentioned at #164 

Right now is needed to manually change the paywall address for testing.

dajohi commented:
@vctt94 fix the test failures?

vctt94 commented:
Tests are failing because of https://github.com/decred/politeiagui/issues/147

sndurkin commented on src/selectors/api.js:
```diff
@@ -138,3 +138,8 @@ export const setStatusProposalIsRequesting = isApiRequestingSetStatusProposal;
 export const redirectedFrom = get(["api", "login", "redirectedFrom"]);
+export const paywallAddress = compose(get("paywalladdress"), apiNewUserResponse);
+export const paywallAmount = compose(get("paywallamount"), apiNewUserResponse);
+export const VerificationToken = compose(get("verificationToken"), apiNewUserResponse);
```
The variable name here should use camelCase, and the string that's sent from Politeia will be all lowercase (same issue with the 2 lines below this).

sndurkin commented on src/connectors/paywall.js:
```diff
@@ -0,0 +1,17 @@
+  sel.selectorMap({
+    paywallAddress: sel.paywallAddress,
+    paywallAmount: sel.paywallAmount,
+    VerificationToken: sel.VerificationToken,
```
We should use camelCase here.

vctt94 commented:
Thanks for reviewing my PR.

I believe we can use in-line instead of a modal, should I remove the modal from the code? The way it is now, I believe it can be reusable for some modal we may have someday. 

I can change the polling for another place where it is always called, like the navbar. But I agree with you, it is better to wait for the server to inform it is paid.

But are we going to wait for the server to say it is paid and check for the mined blocks on the gui? Or will it be all on the server side and the gui only asks if it good to access the submit proposal area?

sndurkin commented:
> But are we going to wait for the server to say it is paid and check for the mined blocks on the gui? Or will it be all on the server side and the gui only asks if it good to access the submit proposal area?

I think the server will reply with `hasPaid` on `/user/login` and `/user/me`, and if it's false, the GUI will poll for mined blocks, because it's possible that the user verifies his email, logs in, and then the transaction is confirmed on the blockchain he clicks around on the site. In that case, we will want to enable commenting and proposal submissions without the user having to press F5 to refresh the browser. Does that make sense?

vctt94 commented:
I removed the modal and made it a normal component, I created the hadPaid, which is get from login and me, and added the poll on the header if it hasPaid === false. But we have a problem, I may not have the paywall address and I am not sure if it makes sense to send it together with `/user/login` and `/user/me` route. What do you think?

sndurkin commented on mocks/api/v1/user/me/GET.json:
```diff
@@ -1,4 +1,5 @@
-  "email": "foo@foo.com",
-  "isadmin": true
-}
+    "hasPaid": true,
```
This should be all lowercase.

jolan commented on src/lib/external_api.js:
```diff
@@ -0,0 +1,14 @@
+  return r.json();
+});
+
+export const getPaymentsByAddress = address => {
```
Makes more sense to just check the first letter of the payment address. i.e. D = mainnet, T = testnet

sndurkin commented on src/components/Paywall/Page.js:
```diff
@@ -0,0 +1,27 @@
+    </div>
+    <div>
+      Status: {
+        grantAccess ?  <p>Approved</p> : <p>Waiting for blocks to get mined</p>
```
Let's call this "Payment received" and "Waiting for payment"

sndurkin commented on src/components/Paywall/Page.js:
```diff
@@ -0,0 +1,27 @@
+      <h2 className="paywall-header-title">Paywall</h2>
+    </div>
+    <div className="paywall-content">
+      <p>Please send exaclty {props.paywallAmount}DCR to <br/>{props.paywallAddress}</p>
```
Typo, should be "exactly".

sndurkin commented on src/components/Paywall/Page.js:
```diff
@@ -0,0 +1,27 @@
+      }
+    </div>
+    <div className="paywall-footer">
+      <a className="btn">Pay!</a>
```
I don't think a button is necessary here, because the user will pay with Decrediton, right?

tiagoalvesdulce commented on src/components/Paywall/index.js:
```diff
@@ -0,0 +1,54 @@
+    this.state = this.getInitialState();
+  }
+
+  getInitialState() {
```
getInitialState should not be used with ES6 classes, you can initialize the state in the constructor.

tiagoalvesdulce commented on src/components/Paywall/index.js:
```diff
@@ -0,0 +1,54 @@
+    this.setState(this.getInitialState());
+  }
+
+  onSubmit() {
```
onSubmit function is empty here.

vctt94 commented:
I fixed your comments and also I added a warningPaywallNotPaid component for not showing the submit proposal button and save button on comments.

sndurkin commented on src/components/snew/CommentForm.js:
```diff
@@ -77,18 +79,23 @@ const CommentForm = ({
-            </div>
+            {
+              !grantAccess ? <WarningPaywallNotPaid
+                message="You can not Comment on proposals before paying the paywall" /> : (
```
2 comments:

* Could you include the Save button but just disable it and the text area? I think it will make it a little clearer that this is the comment form if the button is present too.
* I'd like to refrain from calling it a paywall, can you rephrase it to: "You must pay the registration fee before you can comment on proposals"

sndurkin commented on src/components/snew/SubmitTextSidebox.js:
```diff
@@ -1,26 +1,30 @@
-  </div>
-) : null;
+  ) : <WarningPaywallNotPaid
+    message="You can not send Proposals before paying The Paywall" />
```
Same comments here:

* Could we show the "Submit a new proposal" button but disable it?
* Can you rephrase the message to: "You must pay the registration fee before you can submit a proposal."

tiagoalvesdulce commented on src/actions/external_api.js:
```diff
@@ -0,0 +1,68 @@
+  for (let voutData of transaction["vout"]) {
+    const addresses = voutData["scriptPubKey"]["addresses"];
+    if (addresses) {
+      addresses.forEach(address => {
```
This line is throwing a [no-loop-func](https://eslint.org/docs/rules/no-loop-func) warning. You should basically create a function outside the `for (let voutData ... )` loop and call it inside the `addresses.forEach`.

vctt94 commented:
Made the sean changes and I changed forEach, because it needs a callback function.

tiagoalvesdulce commented on mocks/api/v1/user/me/GET.json:
```diff
@@ -1,5 +1,4 @@
 {
-    "email": "foo@foo.com",
```
Removing this messes up with `yarn uidev` and Travis

vctt94 commented:
Ok, I am redirecting now and I returned the email to the mock api.

I believe we should not have tests depending on the mock API. If the test need to be logged in to pass, it needs to login. Maybe we should have an issue for that.
